client/src/game/ui/selection/shapecontext.vue

```diff
@@ -19,85 +19,139 @@
+        <li v-if="hasSingleShape()" @click="openEditDialog">Show properties</li>
     </ContextMenu>
 </template>
 
 <script lang="ts">
 import Vue from "vue";
 import Component from "vue-class-component";
 
+import { mapState } from "vuex";
+
 import ContextMenu from "@/core/components/contextmenu.vue";
 
+import { socket } from "@/game/api/socket";
+import { ServerLocation } from "@/game/comm/types/general";
 import { EventBus } from "@/game/event-bus";
 import { layerManager } from "@/game/layers/manager";
 import { Shape } from "@/game/shapes/shape";
 import { gameStore } from "@/game/store";
-import { inInitiative, initiativeStore } from "../initiative/store";
+import { cutShapes, deleteShapes, pasteShapes } from "../../shapes/utils";
+import { initiativeStore, inInitiative } from "../initiative/store";
 
 @Component({
     components: {
         ContextMenu,
     },
+    computed: {
+        ...mapState("game", ["locations", "assets", "notes"]),
+    },
 })
 export default class ShapeContext extends Vue {
     visible = false;
     x = 0;
     y = 0;
-    shape: Shape | null = null;
+    shapes: Shape[] | null = null;
     get activeLayer(): string {
         const layer = layerManager.getLayer();
         return layer === undefined ? "" : layer.name;
     }
-    open(event: MouseEvent, shape: Shape) {
+    open(event: MouseEvent, shapes: Shape[]) {
         this.visible = true;
         this.x = event.pageX;
         this.y = event.pageY;
-        this.shape = shape;
+        this.shapes = shapes;
         this.$nextTick(() => (<HTMLElement>this.$children[0].$el).focus());
     }
     close() {
         this.visible = false;
-        this.shape = null;
+        this.shapes = null;
     }
     getLayers() {
         return layerManager.layers.filter(l => l.selectable && (gameStore.IS_DM || l.playerEditable));
     }
     getActiveLayer() {
         return layerManager.getLayer();
     }
+    getCurrentLocation() {
+        return gameStore.locationName;
+    }
     getInitiativeWord() {
-        if (this.shape === null) return "";
-        return inInitiative(this.shape.uuid) ? "Show" : "Add";
+        if (this.shapes === null) return "";
+        if (this.shapes.length === 1) {
+            return inInitiative(this.shapes[0].uuid) ? "Show" : "Add";
+        } else {
+            return this.shapes.every(shape => inInitiative(shape.uuid)) ? "Show" : "Add all to";
+        }
+    }
+    hasSingleShape(): boolean {
+        return this.shapes !== null && this.shapes.length === 1;
     }
     setLayer(newLayer: string) {
-        if (this.shape === null) return;
-        this.shape.moveLayer(newLayer, true);
+        if (this.shapes === null) return;
+        this.shapes.forEach(shape => shape.moveLayer(newLayer, true));
+        this.close();
+    }
+    setLocation(newLocation: string) {
+        if (this.shapes === null) return;
+        const layer = layerManager.getLayer()!;
+        layer.selection = this.shapes;
```

Comment:
Right, I was wondering why you reasign the layer selection, I was under the impression this would already equal this.shapes at this instance, I might be wrong though.

Reply:
Just in case. Probably the correct solution would be to remove this.shapes and just use the layer.selection array and not have the two diferent copies.